### PR TITLE
Update orthoing code to handle SWIR and CAVIS

### PR DESCRIPTION
**Pull request by clairecporter**
While making the changes needed to use 14 bits of input data for SWIR and CAVIS data, I took the opportunity to refactor some of the messy internals of ortho_functions.py and the godawful tests.  This includes

1. Aligning the metadata reading structure and eliminating reading the text files twice
2. Making the ImageInfo class into a proper class instead of an empty bucket to throw things in
3. Adding checks for incompatible arguments
4. Some reformating of un-pythonic function names

Sadly, some of the tests may be beyond rescue, but I have left those I could not repair in the repo for now.

The recent [NoData fix branch](https://github.com/PolarGeospatialCenter/imagery_utils/pull/74) is also rolled in.

### CODE REVIEW

**Review by bakkerbakker — APPROVED** [review]
Looks good, thanks for updating the tests!

**Comment by bakkerbakker** [comment]
@bakkerbakker once this PR gets tested and merged, and Ali's auto DEM functionality gets merged, make a release for imagery_utils

**Comment by clairecporter** [comment]
@power720 I updated the README with info on how to run tests. At PGC the tests are in data/common/repos/repo_testdata/imagery_utils.  Let me know if you feel we should add more to the readme.  I think we should convert the old tests to pytest and ditch the unusable ones at some point.  But that is for another PR.

File: lib/ortho_functions.py [hunk 1]
```diff
@@ -238,9 +204,381 @@
...
+                    rc = 1
+                else:
+                    #### Transform geometries to geographic
+                    if not s_srs.IsSame(g_srs):
+                        ul_geom.Transform(sg_ct)
+                        ur_geom.Transform(sg_ct)
+                        ll_geom.Transform(sg_ct)
+                        lr_geom.Transform(sg_ct)
+                        image_geom.Transform(sg_ct)
+                    logger.info("Geographic extent: %s", str(image_geom))
```
Comment (power720) [inline lib/ortho_functions.py:441]: Log fires on every scene. Consider changing to `logger.debug()` for opt-in verbosity.

File: lib/ortho_functions.py [hunk 2]
```diff
@@ -238,9 +204,381 @@
...
+
+                    #### Get geographic Envelope
+                    self.minlon, self.maxlon, self.minlat, self.maxlat = image_geom.GetEnvelope()
+
+                    ## Determine output image projection if applicable
+                    if type(args.epsg) is str:
+                        self.cent_lat = (self.minlat + self.maxlat) / 2
+                        self.cent_lon = (self.minlon + self.maxlon) / 2
+                        self.epsg = get_epsg_from_lat_lon(self.cent_lat, self.cent_lon, mode=args.epsg, utm_nad83=args.epsg_utm_nad83)
+                        logger.info("Automatically selected output projection EPSG code: %d", self.epsg)
```
Comment (power720) [inline lib/ortho_functions.py:451]: Log fires on every scene. Consider changing to `logger.debug()` for opt-in verbosity.

File: lib/ortho_functions.py [hunk 3]
```diff
@@ -238,9 +204,381 @@
...
+                    gt_ct = osr.CoordinateTransformation(g_srs, t_srs)
+
+                    #### Transform geoms to target srs
+                    if not g_srs.IsSame(t_srs):
+                        ul_geom.Transform(gt_ct)
+                        ur_geom.Transform(gt_ct)
+                        ll_geom.Transform(gt_ct)
+                        lr_geom.Transform(gt_ct)
+                        image_geom.Transform(gt_ct)
+                    logger.info("Projected extent: %s", str(image_geom))
```
Comment (power720) [inline lib/ortho_functions.py:472]: Log fires on every scene. Consider changing to `logger.debug()` for opt-in verbosity.

File: lib/ortho_functions.py [hunk 4]
```diff
@@ -238,9 +204,381 @@
...
+
+                    ####  Make a string for Pixel Size Specification
+                    if args.resolution is not None:
+                        self.res = "-tr {} {} ".format(args.resolution, args.resolution)
+                    else:
+                        self.res = "-tr {0:.12f} {1:.12f} ".format(resx, resy)
+                    if args.tap:
+                        self.tap = "-tap "
+
+                    logger.info("Original image size: %f x %f, res: %.12f x %.12f", rasterxsize_m, rasterysize_m, resx,
```
Comment (power720) [inline lib/ortho_functions.py:495]: Log fires on every scene. Consider changing to `logger.debug()` for opt-in verbosity.

**Comment by power720** [comment]
## Outputs from pgc_ortho.py appear unchanged.
**How I checked:**
I ran a pgc_ortho.py on a batch of images using code from this PR and the same batch using the `common/repos` code. Once created, I `diff`'d the outputs from `gdalinfo` from the corresponding images. 

All of the diffs were some variations of the example below. Differences appear limited to insignificant floating-point values. Manual inspection of a few output images in QGIS also indicated that the outputs were unchanged.

```
2c2
< Files: /mnt/pgc/data/scratch/devin/ortho-pr/ortho-common-repos/GE01_20090705215213_1050410001E50F00_09JUL05215213-P1BS-507561715020_01_P001_u16rf32606.tif
---
> Files: /mnt/pgc/data/scratch/devin/ortho-pr/ortho-swir-bit-depth/GE01_20090705215213_1050410001E50F00_09JUL05215213-P1BS-507561715020_01_P001_u16rf32606.tif
44,45c44,45
< Origin = (422035.505074626125861,7782720.200997613370419)
< Pixel Size = (0.413395693300000,-0.416917544832000)
---
> Origin = (422035.505074621294625,7782720.200997664593160)
> Pixel Size = (0.413395693301000,-0.416917544833000)
181c181
<     STATISTICS_MEAN=206.22784829604
---
>     STATISTICS_MEAN=206.22784832118
183c183
<     STATISTICS_STDDEV=31.592251872396
---
>     STATISTICS_STDDEV=31.592251819676

```

## New logs are much more verbose

The logging messages I commented on create a lot of more standard output than in the previous version of the script. Should this extra output be off by default and available to opt-in via the `--verbose` flag? 

Full standard output from previous version:
```
$ bash pgc_ortho_common_repos.sh
/home/power720/miniforge3/envs/pgc/lib/python3.11/site-packages/osgeo/osr.py:385: FutureWarning: Neither osr.UseExceptions() nor osr.DontUseExceptions() has been explicitly called. In GDAL 4.0, exceptions will be enabled by default.
  warnings.warn(
04-16-2024 16:01:30 INFO- Running command: /mnt/pgc/data/common/repos/imagery_utils/pgc_ortho.py --format GTiff --outtype UInt16 --epsg=utm --dem /mnt/pgc/data/elev/dem/ned/alaskaNED/alaskaned_mosaic_vwgs84.tif --stretch rf --resample near --tasks-per-job 100 --slurm /home/power720/scratch/ortho-pr/renamed /home/power720/scratch/ortho-pr/ortho-common-repos
04-16-2024 16:01:30 INFO- Processing command: /mnt/pgc/data/common/repos/imagery_utils/pgc_ortho.py --format GTiff --outtype UInt16 --epsg=utm --dem /mnt/pgc/data/elev/dem/ned/alaskaNED/alaskaned_mosaic_vwgs84.tif --stretch rf --resample near --tasks-per-job 100 --slurm /home/power720/scratch/ortho-pr/renamed /home/power720/scratch/ortho-pr/ortho-common-repos
04-16-2024 16:01:31 INFO- Number of src images: 38
04-16-2024 16:01:33 INFO- Number of incomplete tasks: 38
Writing task bundle text files in directory: /home/power720/scratch/task_bundles
04-16-2024 16:01:33 INFO- Submitting Tasks
Submitted batch job 1219050
04-16-2024 16:01:33 INFO- Done
```

Trimmed standard output from PR:
```
$ bash pgc_ortho_swir_bit_depth.sh
04-16-2024 16:32:26 INFO- Running command: /home/power720/scratch/ortho-pr/imagery_utils/pgc_ortho.py --format GTiff --outtype UInt16 --epsg=utm --dem /mnt/pgc/data/elev/dem/ned/alaskaNED/alaskaned_mosaic_vwgs84.tif --stretch rf --resample near --tasks-per-job 100 --slurm /home/power720/scratch/ortho-pr/renamed /home/power720/scratch/ortho-pr/ortho-swir-bit-depth
04-16-2024 16:32:26 INFO- Processing command: /home/power720/scratch/ortho-pr/imagery_utils/pgc_ortho.py --format GTiff --outtype UInt16 --epsg=utm --dem /mnt/pgc/data/elev/dem/ned/alaskaNED/alaskaned_mosaic_vwgs84.tif --stretch rf --resample near --tasks-per-job 100 --slurm /home/power720/scratch/ortho-pr/renamed /home/power720/scratch/ortho-pr/ortho-swir-bit-depth
04-16-2024 16:32:26 INFO- Number of src images: 38
04-16-2024 16:32:26 INFO- Geographic extent: POLYGON Z ((-149.041111111111 70.1394444444445 0,-148.628055555556 70.1402777777778 0,-148.627777777778 69.9844444444444 0,-149.041111111111 69.9825 0,-149.041111111111 70.1394444444445 0))
04-16-2024 16:32:26 INFO- Automatically selected output projection EPSG code: 32606
04-16-2024 16:32:26 INFO- Projected extent: POLYGON Z ((422621.591274233 7782720.20099766 0,438279.334519624 7782341.48383028 0,437825.755827772 7764968.33373176 0,422035.505074621 7765226.56673174 0,422621.591274233 7782720.20099766 0))
04-16-2024 16:32:26 INFO- Original image size: 15662.322632 x 17503.449285, res: 1.529673076676 x 1.554066348641
04-16-2024 16:32:26 INFO- Geographic extent: POLYGON Z ((-149.038738521025 68.2403487243692 0,-148.599328719604 68.2367892215348 0,-148.597703543621 68.083735579242 0,-149.04073376696 68.0900191861167 0,-149.038738521025 68.2403487243692 0))
04-16-2024 16:32:26 INFO- Automatically selected output projection EPSG code: 32606
04-16-2024 16:32:26 INFO- Projected extent: POLYGON Z ((415668.677663328 7571055.08411419 0,433830.445491495 7570122.33000709 0,433455.988476124 7553060.93281567 0,415032.041568806 7554304.75990041 0,415668.677663328 7571055.08411419 0))
04-16-2024 16:32:26 INFO- Original image size: 18185.704299 x 16762.418286, res: 0.516947733008 x 0.574547327700
...
04-16-2024 16:32:28 INFO- Geographic extent: POLYGON Z ((-149.030105116295 69.075920977549 0,-148.596194255114 69.0704927328091 0,-148.597167624743 68.9910969066389 0,-149.031187048465 68.9965849112837 0,-149.030105116295 69.075920977549 0))
04-16-2024 16:32:28 INFO- Automatically selected output projection EPSG code: 32606
04-16-2024 16:32:28 INFO- Projected extent: POLYGON Z ((419100.226717474 7664166.59184101 0,436372.057123829 7663050.32058574 0,436102.8292286 7654200.86479256 0,418764.286570471 7655325.78864324 0,419100.226717474 7664166.59184101 0))
04-16-2024 16:32:28 INFO- Original image size: 17307.864891 x 8847.183561, res: 2.004152951768 x 2.062280550377
04-16-2024 16:32:28 INFO- Number of incomplete tasks: 38
Writing task bundle text files in directory: /home/power720/scratch/task_bundles
04-16-2024 16:32:28 INFO- Submitting Tasks
Submitted batch job 1219085
04-16-2024 16:32:28 INFO- Done
```

File: README.md [hunk 5]
```diff
@@ -37,30 +37,50 @@ The mosaicking toolset mosaics multiple input images into a set of non-overlappi
...
-https://github.com/conda-forge/miniforge#mambaforge
+## Installation and dependencies
+PGC uses the Miniforge installer to build our Python/GDAL software stack.  You can find installers for your OS here:
+https://github.com/conda-forge/miniforge?tab=readme-ov-file#miniforge3
 
 Users should expect a recent (less than 1-2 years old) version of Python and GDAL to be compatible with tools in this repo.
 The following conda/mamba environment likely contains more dependencies than are needed for tools in this repo, but should suffice:
 ```
-conda create --name pgc -c conda-forge python=3.11 gdal=3.6.4 numpy scipy pandas geopandas rasterio shapely postgresql psycopg2 sqlalchemy configargparse lxml pathlib2 python-dateutil pytest rtree xlsxwriter
+mamba create --name pgc -c conda-forge git python=3.11 gdal=3.6.4 globus-sdk globus-cli numpy scipy pandas geopandas rasterio shapely postgresql psycopg2 sqlalchemy configargparse lxml pathlib2 python-dateutil pytest rtree xlsxwriter tqdm alive-progress pyperclip --yes
```
Comment (power720) [inline README.md:60]: To make this match our current `pgc` environment the following additional changes would be needed:

Bump: `gdal=3.7.2`
Add: `dbfread openpyxl pyperclip xlsxwriter`
Remove: None.

**Comment by power720** [comment]
I was able to get the tests setup and running.

I think I've reached the limit of my relevant repo knowledge, so I do not have any further comments. Let me know if you'd like to discuss them or if you have any more specific items that you'd like me to review closely.

File: README.md (hunk 5 above)
Comment (clairecporter) [reply README.md:60]: Good catch!

**Comment by clairecporter** [comment]
The logging issue is because a lot of the work that used to be done by several different functions (all with different logging practices) and now they are all in a single class.  When the class is instantiated as you did for submission to Rookery, the logging messages are overkill, but when the class is instantiated later on in the code within the same run (if not submitting to slurm) or within the slurm job, the log messages are expected even without `--verbose` set.  I'll have to think about how to change the info level of logging depending on the context.

**Comment by clairecporter** [comment]
@power720, I simplified the logging for `pgc_ortho.py` to make it more useful.  I also found an inefficiency in how the ImageInfo class was being used that I ironed out.  I'll be merging this in shortly.

**Pull request merged**